File: tests/rosetta/transpiler/FS/cantor-set.fs
// Generated 2025-07-27 16:40 +0000

exception Break
exception Continue

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

type Anon1 = {
    start: int
    len: int
    index: int
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let width: int = 81
let height: int = 5
let mutable lines: string array = [||]
for i in 0 .. (height - 1) do
    let mutable row: string = ""
    let mutable j: int = 0
    while j < width do
        row <- row + "*"
        j <- j + 1
    lines <- unbox<string array> (Array.append lines [|row|])
let rec setChar (s: string) (idx: int) (ch: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    let mutable idx = idx
    let mutable ch = ch
    try
        __ret <- ((_substring s 0 idx) + ch) + (_substring s (idx + 1) (String.length s))
        raise Return
        __ret
    with
        | Return -> __ret
let mutable stack: Map<string, int> array = [|Map.ofList [("start", 0); ("len", width); ("index", 1)]|]
try
    while (unbox<int> (Array.length stack)) > 0 do
        let mutable frame: Map<string, int> = stack.[(unbox<int> (Array.length stack)) - 1]
        stack <- unbox<Map<string, int> array> (Array.sub stack 0 ((unbox<int> ((unbox<int> (Array.length stack)) - 1)) - 0))
        let start: int = frame.["start"] |> unbox<int>
        let lenSeg: int = frame.["len"] |> unbox<int>
        let index: int = frame.["index"] |> unbox<int>
        let seg: int = unbox<int> (lenSeg / 3)
        if seg = 0 then
            raise Continue
        let mutable i: int = index
        while i < height do
            let mutable j: int = start + seg
            while j < (start + (2 * seg)) do
                lines.[i] <- setChar (unbox<string> (lines.[i])) j " "
                j <- j + 1
            i <- i + 1
        stack <- unbox<Map<string, int> array> (Array.append stack [|Map.ofList [("start", start); ("len", seg); ("index", index + 1)]|])
        stack <- unbox<Map<string, int> array> (Array.append stack [|Map.ofList [("start", start + (seg * 2)); ("len", seg); ("index", index + 1)]|])
with
| Break -> ()
| Continue -> ()
for line in lines do
    printfn "%A" line
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/rosetta/transpiler/php/cantor-set.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  $width = 81;
  $height = 5;
  $lines = [];
  for ($i = 0; $i < $height; $i++) {
  $row = '';
  $j = 0;
  while ($j < $width) {
  $row = $row . '*';
  $j = $j + 1;
};
  $lines = array_merge($lines, [$row]);
}
  function setChar($s, $idx, $ch) {
  global $width, $height, $lines, $row, $j, $stack, $frame, $start, $lenSeg, $index, $seg, $i;
  return substr($s, 0, $idx - 0) . $ch . substr($s, $idx + 1, strlen($s) - ($idx + 1));
};
  $stack = [['start' => 0, 'len' => $width, 'index' => 1]];
  while (count($stack) > 0) {
  $frame = $stack[count($stack) - 1];
  $stack = array_slice($stack, 0, count($stack) - 1 - 0);
  $start = $frame['start'];
  $lenSeg = $frame['len'];
  $index = $frame['index'];
  $seg = intval((_intdiv($lenSeg, 3)));
  if ($seg == 0) {
  continue;
}
  $i = $index;
  while ($i < $height) {
  $j = $start + $seg;
  while ($j < $start + 2 * $seg) {
  $lines[$i] = setChar($lines[$i], $j, ' ');
  $j = $j + 1;
};
  $i = $i + 1;
};
  $stack = array_merge($stack, [['start' => $start, 'len' => $seg, 'index' => $index + 1]]);
  $stack = array_merge($stack, [['start' => $start + $seg * 2, 'len' => $seg, 'index' => $index + 1]]);
}
  foreach ($lines as $line) {
  echo rtrim($line), PHP_EOL;
}
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = intdiv($__end - $__start, 1000);
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;;
